editoast/editoast_derive/src/lib.rs

```diff
@@ -237,3 +237,37 @@ pub fn model(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
         .unwrap_or_else(darling::Error::write_errors)
         .into()
 }
+
+#[cfg(test)]
+pub(crate) fn pretty_tokens(tokens: &proc_macro2::TokenStream) -> String {
+    let file = syn::parse_file(tokens.to_string().as_str()).unwrap();
+    prettyplease::unparse(&file)
+}
+
+// Extra allows because of some false negative lints, not sure why.
+// Maybe because editoast_derive is a proc-macro crate?
+
+#[cfg(test)]
+#[allow(unused_macros)]
+macro_rules! assert_macro_expansion {
```

Comment:
Why do we need a macro here? Isn't it a function that would take a `Fn` and a `DeriveInput` as input arguments?

Reply:
It's to keep the right failure source location. If we put the assert in a function, should it fail, the top stack frame would indicate that function, instead of the expected location in our failing test.

I don't have a strong preference for the macro, but I find it a bit cleaner, lmk what you prefer.

Reply:
I didn't know about this one thanks! 😮

Seems like what I want :)

Reply:
It works for the assertion location, but `insta` does some magic to retrieve the name of the test, which basically resolves to getting the function name. From the expansion of `assert_snapshot!`:

```rust
            fn f(){}
            
            fn type_name_of_val<T>(_:T) ->  &'static str {
                std::any::type_name::<T>()
            }
            let mut name = type_name_of_val(f).strip_suffix("::f").unwrap_or("");
            while let Some(rest) = name.strip_suffix("::{{closure}}"){
                name = rest;
            }name
```

Placing the `assert` in a function results with all the snapshots being suffixed by `assert_macro_expansion` instead of the test name:

![image](https://github.com/user-attachments/assets/8635996c-374c-48be-bb62-b93454c139e9)

I guess it's better to keep the macro under these conditions 🤷

Reply:
I see, berk. I've looked a bit, and `assert_snapshot!()` does call `_assert_snapshot_base!()` (exposed but `#[doc(hidden)]`) which itself calls the function `insta::_macro_support::assert_snapshot` (the entire module `_macro_support` is `pub` but also `#[doc(hidden)]`). So it might be possible... but I don't have the time to dig that rabbit hole. I'm resolving.